vmm-reference-gdt.md

```diff
@@ -0,0 +1,185 @@
+This module provides abstractions for building a Global Descriptor Table (GDT).
+
+## Segment Descriptors
+Segment Descriptors for the GDT have been defined as follows:
+```rs
+pub struct SegmentDescriptor(pub u64);
+```
+
+The following function creates a `SegmentDescriptor` given base, limit and flags.
+The way to convert these numbers to a 64 - bit number has been described [here](https://wiki.osdev.org/Global_Descriptor_Table#Segment_Descriptor). 
+
+`base` is a 32 bit value and specifies the address where this segment starts.
+
+`limit` is a 32 bit value, of which only the least significant 20 bits are used. It tells the maximum addressable unit.
+
+`flags` is a 16 bit value, of which 12 are used. It has two parts. The last 8 bits are referred to as Access Byte. The first 4 bits are some more flags.
+
+```rs
+    pub fn from(flags: u16, base: u32, limit: u32) -> Self;
+```
+<table class="wikitable">
+<caption> Segment Descriptor
+</caption>
+<tbody><tr>
+<th style="width: 20%; text-align: left;">63&nbsp;&nbsp;&nbsp;<span style="float: right;">56</span>
+</th>
+<th style="width: 12.5%; text-align: left;">55&nbsp;&nbsp;&nbsp;<span style="float: right;">52</span>
+</th>
+<th style="width: 12.5%; text-align: left;">51&nbsp;&nbsp;&nbsp;<span style="float: right;">48</span>
+</th>
+<th style="width: 25%; text-align: left;">47&nbsp;&nbsp;&nbsp;<span style="float: right;">40</span>
+</th>
+<th style="width: 25%; text-align: left;">39&nbsp;&nbsp;&nbsp;<span style="float: right;">32</span>
+</th></tr>
+<tr>
+<td><b>Base</b><br>31&nbsp;&nbsp;&nbsp;<span style="float: right;">24</span>
+</td>
+<td><b>Flags</b><br>3&nbsp;&nbsp;&nbsp;<span style="float: right;">0</span>
+</td>
+<td><b>Limit</b><br>19&nbsp;&nbsp;&nbsp;<span style="float: right;">16</span>
+</td>
+<td><b>Access Byte</b><br>7&nbsp;&nbsp;&nbsp;<span style="float: right;">0</span>
+</td>
+<td><b>Base</b><br>23&nbsp;&nbsp;&nbsp;<span style="float: right;">16</span>
+</td></tr>
+<tr>
+<th colspan="3" style="text-align: left;">31 &nbsp;&nbsp;<span style="float: right;">16</span>
+</th>
+<th colspan="2" style="text-align: left;">15 &nbsp;&nbsp;<span style="float: right;">0</span>
+</th></tr>
+<tr>
+<td colspan="3"><b>Base</b><br>15&nbsp;&nbsp;&nbsp;<span style="float: right;">0</span>
+</td>
+<td colspan="2"><b>Limit</b><br>15&nbsp;&nbsp;&nbsp;<span style="float: right;">0</span>
+</td></tr></tbody><div></div></table>
+
+The following methods extract base and limit from a `SegmentDescriptor`.
+```rs
+    fn base(&self) -> u64;
+    fn limit(&self) -> u32;
+```
+These extract some flags from the segment.
+```rs
+    fn g(&self) -> u8;
+    fn db(&self) -> u8;
+    fn l(&self) -> u8;
+    fn avl(&self) -> u8;
+```
+The following values are extracted from the access byte poriton of the flags. Description available [here](https://wiki.osdev.org/Global_Descriptor_Table#Segment_Descriptor).
+```rs
+    fn p(&self) -> u8;
+    fn dpl(&self) -> u8;
+    fn s(&self) -> u8;
```

Comment:
Would be good to give one line description in-place of what each bit is. Instead of making us click the link.